clientgenv2/source_generator.go

```diff
@@ -76,6 +70,114 @@ func (rs ResponseFieldList) IsStructType() bool {
 	return len(rs) > 0 && !rs.IsFragment()
 }
 
+func (rs ResponseFieldList) MapByName() map[string]*ResponseField {
+	res := make(map[string]*ResponseField)
+	for _, field := range rs {
+		res[field.Name] = field
+	}
+	return res
+}
+
+func (rs ResponseFieldList) SortByName() ResponseFieldList {
+	sort.Slice(rs, func(i, j int) bool {
+		return rs[i].Name < rs[j].Name
+	})
+	return rs
+}
+
+type StructGenerator struct {
+	currentResponseFieldList ResponseFieldList
+	preMergedStructSources   []*StructSource
+	postMergedStructSources  []*StructSource
+}
+
+func NewStructGenerator(responseFieldList ResponseFieldList) *StructGenerator {
+	currentFields := make(ResponseFieldList, 0)
+	fragmentChildrenFields := make(ResponseFieldList, 0)
+	for _, field := range responseFieldList {
+		if field.IsFragmentSpread {
+			fragmentChildrenFields = append(fragmentChildrenFields, field.ResponseFields...)
+		} else {
+			currentFields = append(currentFields, field)
+		}
+	}
+
+	preMergedStructSources := make([]*StructSource, 0)
+	postMergedStructSources := make([]*StructSource, 0)
+
+	for _, field := range responseFieldList {
+		if field.IsFragmentSpread {
+			preMergedStructSources = append(preMergedStructSources, &StructSource{
+				Name: field.FieldTypeString(),
+				Type: field.ResponseFields.StructType(),
+			})
+		}
+	}
+
+	currentFields = mergeFieldsRecursively(currentFields, fragmentChildrenFields, &preMergedStructSources, &postMergedStructSources)
+	return &StructGenerator{
+		currentResponseFieldList: currentFields,
+		preMergedStructSources:   preMergedStructSources,
+		postMergedStructSources:  postMergedStructSources,
+	}
+}
+
+func mergeFieldsRecursively(targetFields ResponseFieldList, sourceFields ResponseFieldList, preMerged, postMerged *[]*StructSource) ResponseFieldList {
```

Comment:
It’s generally not a good practice to implicitly cause side effects by taking a pointer as an argument and modifying its value. I think it would be better to return the value explicitly as a return value.

```suggestion
func mergeFieldsRecursively(targetFields ResponseFieldList, sourceFields ResponseFieldList, preMerged []*StructSource) (ResponseFieldList, []*StructSource, []*StructSource) {
```

Reply:
Okey, I will follow you.